### ISSUE TITLE
Expand performance section of the guide

### ISSUE THREAD

**Issue by adamreichold** [PyO3/pyo3#3310]
- [ ] Overhead of conversions (i.e. `PyList` vs `Vec` etc.) 
- [ ] String `intern!`
- [ ] `Vec<u8>` becoming `list[int]` (and thus being very slow), usage of `Cow<[u8]>` as an alternative
- [ ] `#[pyo3(get)]` deep-cloning non-`Py` data
- [ ] dictionary dispatch, i.e. look-up based on type object identity

**Comment by davidhewitt** [comment]
Error creation e.g. `PyErrArguments`, how to avoid lazy error construction etc.

**Comment by samuelcolvin** [comment]
Also `.extract()` is massively slower when it fails than `.downcast()`

**Comment by davidhewitt** [comment]
@samuelcolvin we added that as the initial content of the performance section already 😉 

https://pyo3.rs/main/performance#extract-versus-downcast

**Comment by Icxolu** [comment]
> * [Its return value is indeed `bytes`](https://github.com/PyO3/pyo3/blob/241080aeba3e97f29ac2617cd89c0e26563d6703/src/conversions/std/num.rs#L267-L277), so there is no problem as mentioned above, right?

Not anymore. We fixed this long standing footgun with the introduction of `IntoPyObject`.

> But for the parameter `a`, this is its [implementation](https://github.com/PyO3/pyo3/blob/241080aeba3e97f29ac2617cd89c0e26563d6703/src/types/sequence.rs#L334-L370), so it is indeed implemented in the following way, right?

Yes, `FromPyObject` still has the problem of treating every `Vec<T>` equally. I looked into a fix for this as part of the ongoing `FromPyObject` rework https://github.com/PyO3/pyo3/pull/4390#issuecomment-2480501756, but the `IntoPyObject` approach is currently not compatible with the direction we planned for `FromPyObject`. So for the time being this will probably stay as is. PR improving documentation around this are of course welcome 😄

**Comment by WSH032** [comment]
> Yes, `FromPyObject` still has the problem of treating every `Vec<T>` equally. I looked into a fix for this as part of the ongoing `FromPyObject` rework [#4390 (comment)](https://github.com/PyO3/pyo3/pull/4390#issuecomment-2480501756), but the `IntoPyObject` approach is currently not compatible with the direction we planned for `FromPyObject`. So for the time being this will probably stay as is. PR improving documentation around this are of course welcome 😄

If I understand correctly, #5244 v0.27 resolves this issue 🎉.

**Comment by Icxolu** [comment]
> If I understand correctly, [#5244](https://github.com/PyO3/pyo3/pull/5244) v0.27 resolves this issue 🎉.

Yes, we found a way to do a similar specialization for `FromPyObject` as well.

**Comment by JMak-Security** [comment]
Opened #6336 for the `intern!` checklist item. Left the `Vec<u8>`/`Cow<[u8]>` one out since @Icxolu's comment above confirms it's already resolved by the `IntoPyObject` rework — documenting it now would be describing a fixed problem as current. The remaining items (general conversion overhead, `#[pyo3(get)]` cloning, dictionary dispatch) need more investigation than one PR, leaving those open.
